File: server/src/main/kotlin/com/example/reconix/server/service/InvoiceService.kt
package com.example.reconix.server.service

import com.example.reconix.server.database.*
import com.example.reconix.shared.*
import kotlinx.datetime.Clock
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.transactions.transaction
import java.util.UUID

/**
 * Invoice Service - Handles 3-Way Match Validation Logic
 *
 * Business Rules:
 * 1. Invoice Quantity must be <= Sum(GRN Quantities) for that item
 * 2. Invoice Unit Price must match PO Unit Price (±$0.05 tolerance)
 */
class InvoiceService {

    companion object {
        private const val PRICE_TOLERANCE = 0.05
    }

    /**
     * Validate an invoice using 3-Way Match logic
     * Compares Invoice against PO and GRNs
     */
    fun validateInvoice(invoice: InvoiceDTO): ValidationResult {
        return transaction {
            val timestamp = Clock.System.now().toString()

            // 1. Fetch the Purchase Order
            val po = PurchaseOrders.selectAll()
                .where { PurchaseOrders.id eq invoice.poId }
                .singleOrNull()
                ?: return@transaction ValidationResult(
                    status = InvoiceStatus.MANUAL_REVIEW,
                    message = "Purchase Order ${invoice.poId} not found - requires manual review",
                    timestamp = timestamp
                )

            // 2. Fetch PO Items
            val poItems = PurchaseOrderItems.selectAll()
                .where { PurchaseOrderItems.poId eq invoice.poId }
                .associate { row ->
                    row[PurchaseOrderItems.itemId] to Pair(
                        row[PurchaseOrderItems.quantity],
                        row[PurchaseOrderItems.unitPrice]
                    )
                }

            // 3. Fetch all GRNs for this PO and sum quantities per item
            val grnIds = Grns.selectAll()
                .where { Grns.poId eq invoice.poId }
                .map { it[Grns.id] }

            val grnQuantities = mutableMapOf<String, Int>()
            for (grnId in grnIds) {
                GrnItems.selectAll()
                    .where { GrnItems.grnId eq grnId }
                    .forEach { row ->
                        val itemId = row[GrnItems.itemId]
                        val qty = row[GrnItems.receivedQuantity]
                        grnQuantities[itemId] = grnQuantities.getOrDefault(itemId, 0) + qty
                    }
            }

            // 4. Validate each invoice item
            val validationErrors = mutableListOf<String>()

            for (invoiceItem in invoice.items) {
                val itemId = invoiceItem.itemId

                // Check if item exists in PO
                val poItemData = poItems[itemId]
                if (poItemData == null) {
                    validationErrors.add("Item $itemId not found in PO ${invoice.poId}")
                    continue
                }

                val (poQuantity, poUnitPrice) = poItemData

                // Rule 1: Invoice Quantity <= Sum(GRN Quantities)
                val totalReceivedQty = grnQuantities[itemId] ?: 0
                if (invoiceItem.quantity > totalReceivedQty) {
                    validationErrors.add(
                        "Item $itemId: Invoice qty (${invoiceItem.quantity}) exceeds received qty ($totalReceivedQty)"
                    )
                }

                // Rule 2: Invoice Unit Price must match PO Unit Price (±$0.05 tolerance)
                val priceDifference = kotlin.math.abs(invoiceItem.unitPrice - poUnitPrice)
                if (priceDifference > PRICE_TOLERANCE) {
                    validationErrors.add(
                        "Item $itemId: Invoice price ($${invoiceItem.unitPrice}) differs from PO price ($${poUnitPrice}) beyond tolerance"
                    )
                }
            }

            // 5. Determine final status
            val finalStatus = if (validationErrors.isEmpty()) {
                InvoiceStatus.MATCHED
            } else {
                InvoiceStatus.MISMATCH
            }

            // 6. Save or update invoice in database
            val existingInvoice = Invoices.selectAll()
                .where { Invoices.id eq invoice.id }
                .singleOrNull()

            if (existingInvoice == null) {
                // Insert new invoice
                Invoices.insert {
                    it[id] = invoice.id
                    it[poId] = invoice.poId
                    it[vendorId] = invoice.vendorId
                    it[totalAmount] = invoice.totalAmount
                    it[status] = finalStatus.name
                    it[createdAt] = timestamp
                    it[validatedAt] = timestamp
                }

                // Insert invoice items
                for (item in invoice.items) {
                    InvoiceItems.insert {
                        it[invoiceId] = invoice.id
                        it[itemId] = item.itemId
                        it[quantity] = item.quantity
                        it[unitPrice] = item.unitPrice
                    }
                }
            } else {
                // Update existing invoice status
                Invoices.update({ Invoices.id eq invoice.id }) {
                    it[status] = finalStatus.name
                    it[validatedAt] = timestamp
                }
            }

            // 7. Return validation result
            val message = if (validationErrors.isEmpty()) {
                "Invoice validated successfully - 3-Way Match PASSED"
            } else {
                "3-Way Match FAILED: ${validationErrors.joinToString("; ")}"
            }

            ValidationResult(
                status = finalStatus,
                message = message,
                timestamp = timestamp
            )
        }
    }

    /**
     * Get all Purchase Orders
     */
    fun getAllPurchaseOrders(): List<PurchaseOrderDTO> {
        return transaction {
            PurchaseOrders.selectAll().map { poRow ->
                val poId = poRow[PurchaseOrders.id]

                val items = PurchaseOrderItems.selectAll()
                    .where { PurchaseOrderItems.poId eq poId }
                    .map { itemRow ->
                        PurchaseOrderItemDTO(
                            itemId = itemRow[PurchaseOrderItems.itemId],
                            itemName = itemRow[PurchaseOrderItems.itemName],
                            quantity = itemRow[PurchaseOrderItems.quantity],
                            unitPrice = itemRow[PurchaseOrderItems.unitPrice]
                        )
                    }

                PurchaseOrderDTO(
                    id = poId,
                    vendorName = poRow[PurchaseOrders.vendorName],
                    totalAmount = poRow[PurchaseOrders.totalAmount],
                    items = items
                )
            }
        }
    }

    /**
     * Get Purchase Order by ID
     */
    fun getPurchaseOrderById(id: String): PurchaseOrderDTO? {
        return transaction {
            val poRow = PurchaseOrders.selectAll()
                .where { PurchaseOrders.id eq id }
                .singleOrNull() ?: return@transaction null

            val items = PurchaseOrderItems.selectAll()
                .where { PurchaseOrderItems.poId eq id }
                .map { itemRow ->
                    PurchaseOrderItemDTO(
                        itemId = itemRow[PurchaseOrderItems.itemId],
                        itemName = itemRow[PurchaseOrderItems.itemName],
                        quantity = itemRow[PurchaseOrderItems.quantity],
                        unitPrice = itemRow[PurchaseOrderItems.unitPrice]
                    )
                }

            PurchaseOrderDTO(
                id = id,
                vendorName = poRow[PurchaseOrders.vendorName],
                totalAmount = poRow[PurchaseOrders.totalAmount],
                items = items
            )
        }
    }

    /**
     * Get all Invoices
     */
    fun getAllInvoices(): List<InvoiceDTO> {
        return transaction {
            Invoices.selectAll().map { invRow ->
                val invoiceId = invRow[Invoices.id]

                val items = InvoiceItems.selectAll()
                    .where { InvoiceItems.invoiceId eq invoiceId }
                    .map { itemRow ->
                        InvoiceItemDTO(
                            itemId = itemRow[InvoiceItems.itemId],
                            quantity = itemRow[InvoiceItems.quantity],
                            unitPrice = itemRow[InvoiceItems.unitPrice]
                        )
                    }

                InvoiceDTO(
                    id = invoiceId,
                    poId = invRow[Invoices.poId],
                    vendorId = invRow[Invoices.vendorId],
                    totalAmount = invRow[Invoices.totalAmount],
                    status = InvoiceStatus.valueOf(invRow[Invoices.status]),
                    items = items
                )
            }
        }
    }

    /**
     * Get GRNs by PO ID
     */
    fun getGrnsByPoId(poId: String): List<GrnDTO> {
        return transaction {
            Grns.selectAll()
                .where { Grns.poId eq poId }
                .map { grnRow ->
                    val grnId = grnRow[Grns.id]

                    val items = GrnItems.selectAll()
                        .where { GrnItems.grnId eq grnId }
                        .map { itemRow ->
                            GrnItemDTO(
                                itemId = itemRow[GrnItems.itemId],
                                receivedQuantity = itemRow[GrnItems.receivedQuantity]
                            )
                        }

                    GrnDTO(
                        id = grnId,
                        poId = poId,
                        items = items,
                        receivedAt = grnRow[Grns.receivedAt]
                    )
                }
        }
    }
}

File: python_server/services/invoice_service.py
"""
services/invoice_service.py - Invoice validation and PO/GRN queries
Equivalent to server/.../service/InvoiceService.kt
"""
from datetime import datetime, timezone
from typing import List, Optional

from sqlalchemy.orm import Session

from models import (
    Grn, GrnItem, Invoice, InvoiceItem,
    PurchaseOrder, PurchaseOrderItem, ValidationLog,
)
from schemas import (
    GrnDTO, GrnItemDTO, InvoiceDTO, InvoiceItemDTO, InvoiceStatus,
    PurchaseOrderDTO, PurchaseOrderItemDTO, ValidationDetailDTO, ValidationResult,
)

PRICE_TOLERANCE = 0.05


def validate_invoice(db: Session, invoice: InvoiceDTO) -> ValidationResult:
    timestamp = datetime.now(timezone.utc).isoformat()

    po = db.query(PurchaseOrder).filter(PurchaseOrder.id == invoice.poId).first()
    if not po:
        return ValidationResult(
            status=InvoiceStatus.MANUAL_REVIEW,
            message=f"Purchase Order {invoice.poId} not found - requires manual review",
            timestamp=timestamp,
        )

    # PO items: {itemId -> (item_name, quantity, unit_price)}
    po_items = {
        row.item_id: (row.item_name, row.quantity, row.unit_price)
        for row in db.query(PurchaseOrderItem).filter(PurchaseOrderItem.po_id == invoice.poId).all()
    }

    # GRN quantities: {itemId -> total_received}
    grn_ids = [g.id for g in db.query(Grn).filter(Grn.po_id == invoice.poId).all()]
    grn_quantities: dict[str, int] = {}
    for grn_id in grn_ids:
        for gi in db.query(GrnItem).filter(GrnItem.grn_id == grn_id).all():
            grn_quantities[gi.item_id] = grn_quantities.get(gi.item_id, 0) + gi.received_quantity

    validation_errors: List[str] = []
    validation_details = []   # collected for ValidationLog
    for item in invoice.items:
        po_data = po_items.get(item.itemId)
        if po_data is None:
            validation_errors.append(f"Item {item.itemId} not found in PO {invoice.poId}")
            continue
        _item_name, _po_qty, po_price = po_data
        received_qty = grn_quantities.get(item.itemId, 0)
        quantity_match = item.quantity <= received_qty
        price_diff = abs(item.unitPrice - po_price)
        price_match = price_diff <= PRICE_TOLERANCE

        if not quantity_match:
            validation_errors.append(
                f"Item {item.itemId}: Invoice qty ({item.quantity}) exceeds received qty ({received_qty})"
            )
        if not price_match:
            validation_errors.append(
                f"Item {item.itemId}: Invoice price (${item.unitPrice}) differs from PO price (${po_price}) beyond tolerance"
            )

        validation_details.append({
            "item_id": item.itemId,
            "po_quantity": _po_qty,
            "grn_quantity": received_qty,
            "invoice_quantity": item.quantity,
            "po_price": po_price,
            "invoice_price": item.unitPrice,
            "price_difference": price_diff,
            "quantity_match": quantity_match,
            "price_match": price_match,
            "overall_match": quantity_match and price_match,
        })

    final_status = InvoiceStatus.MATCHED if not validation_errors else InvoiceStatus.MISMATCH

    # Upsert invoice record
    existing = db.query(Invoice).filter(Invoice.id == invoice.id).first()
    if existing is None:
        db.add(Invoice(
            id=invoice.id,
            po_id=invoice.poId,
            vendor_id=invoice.vendorId,
            total_amount=invoice.totalAmount,
            status=final_status.value,
            created_at=timestamp,
            validated_at=timestamp,
        ))
        for it in invoice.items:
            db.add(InvoiceItem(
                invoice_id=invoice.id,
                item_id=it.itemId,
                quantity=it.quantity,
                unit_price=it.unitPrice,
            ))
    else:
        existing.status = final_status.value
        existing.validated_at = timestamp

    # Persist ValidationLog for each line item
    for det in validation_details:
        db.add(ValidationLog(
            invoice_id=invoice.id,
            item_id=det["item_id"],
            po_quantity=det["po_quantity"],
            grn_quantity=det["grn_quantity"],
            invoice_quantity=det["invoice_quantity"],
            po_price=det["po_price"],
            invoice_price=det["invoice_price"],
            price_difference=det["price_difference"],
            quantity_match=det["quantity_match"],
            price_match=det["price_match"],
            overall_match=det["overall_match"],
            timestamp=timestamp,
        ))

    db.commit()

    message = (
        "Invoice validated successfully - 3-Way Match PASSED"
        if not validation_errors
        else "3-Way Match FAILED: " + "; ".join(validation_errors)
    )

    # Build ValidationDetail list for the response
    details = [
        ValidationDetailDTO(
            itemId=det["item_id"],
            itemName=po_items[det["item_id"]][0] if det["item_id"] in po_items else det["item_id"],
            poQuantity=det["po_quantity"],
            grnQuantity=det["grn_quantity"],
            invoiceQuantity=det["invoice_quantity"],
            poPrice=det["po_price"],
            invoicePrice=det["invoice_price"],
            priceDifference=det["price_difference"],
            quantityMatch=det["quantity_match"],
            priceMatch=det["price_match"],
            overallMatch=det["overall_match"],
        )
        for det in validation_details
    ]
    return ValidationResult(status=final_status, message=message, timestamp=timestamp, details=details)


def get_all_purchase_orders(db: Session) -> List[PurchaseOrderDTO]:
    result = []
    for po in db.query(PurchaseOrder).all():
        items = [
            PurchaseOrderItemDTO(
                itemId=i.item_id,
                itemName=i.item_name,
                quantity=i.quantity,
                unitPrice=i.unit_price,
            )
            for i in db.query(PurchaseOrderItem).filter(PurchaseOrderItem.po_id == po.id).all()
        ]
        result.append(PurchaseOrderDTO(
            id=po.id,
            vendorName=po.vendor_name,
            vendorEmail=po.vendor_email or "",
            totalAmount=po.total_amount,
            items=items,
        ))
    return result


def get_purchase_order_by_id(db: Session, po_id: str) -> Optional[PurchaseOrderDTO]:
    po = db.query(PurchaseOrder).filter(PurchaseOrder.id == po_id).first()
    if not po:
        return None
    items = [
        PurchaseOrderItemDTO(
            itemId=i.item_id,
            itemName=i.item_name,
            quantity=i.quantity,
            unitPrice=i.unit_price,
        )
        for i in db.query(PurchaseOrderItem).filter(PurchaseOrderItem.po_id == po_id).all()
    ]
    return PurchaseOrderDTO(
        id=po.id,
        vendorName=po.vendor_name,
        vendorEmail=po.vendor_email or "",
        totalAmount=po.total_amount,
        items=items,
    )


def get_all_invoices(db: Session) -> List[InvoiceDTO]:
    result = []
    for inv in db.query(Invoice).all():
        items = [
            InvoiceItemDTO(itemId=it.item_id, quantity=it.quantity, unitPrice=it.unit_price)
            for it in db.query(InvoiceItem).filter(InvoiceItem.invoice_id == inv.id).all()
        ]
        result.append(InvoiceDTO(
            id=inv.id,
            poId=inv.po_id or "",
            vendorId=inv.vendor_id,
            totalAmount=inv.total_amount,
            status=InvoiceStatus(inv.status),
            items=items,
            rejectionReason=inv.rejection_reason,
        ))
    return result


def get_invoices_by_vendor(db: Session, vendor_id: str) -> List[InvoiceDTO]:
    """Return only invoices belonging to the given vendor_id."""
    result = []
    for inv in db.query(Invoice).filter(Invoice.vendor_id == vendor_id).all():
        items = [
            InvoiceItemDTO(itemId=it.item_id, quantity=it.quantity, unitPrice=it.unit_price)
            for it in db.query(InvoiceItem).filter(InvoiceItem.invoice_id == inv.id).all()
        ]
        result.append(InvoiceDTO(
            id=inv.id,
            poId=inv.po_id or "",
            vendorId=inv.vendor_id,
            totalAmount=inv.total_amount,
            status=InvoiceStatus(inv.status),
            items=items,
            rejectionReason=inv.rejection_reason,
        ))
    return result


def get_invoice_by_id(db: Session, invoice_id: str) -> Optional[InvoiceDTO]:
    inv = db.query(Invoice).filter(Invoice.id == invoice_id).first()
    if not inv:
        return None
    items = [
        InvoiceItemDTO(itemId=it.item_id, quantity=it.quantity, unitPrice=it.unit_price)
        for it in db.query(InvoiceItem).filter(InvoiceItem.invoice_id == inv.id).all()
    ]
    return InvoiceDTO(
        id=inv.id,
        poId=inv.po_id or "",
        vendorId=inv.vendor_id,
        totalAmount=inv.total_amount,
        status=InvoiceStatus(inv.status),
        items=items,
        rejectionReason=inv.rejection_reason,
    )


def get_grns_by_po_id(db: Session, po_id: str) -> List[GrnDTO]:
    result = []
    for grn in db.query(Grn).filter(Grn.po_id == po_id).all():
        items = [
            GrnItemDTO(itemId=gi.item_id, receivedQuantity=gi.received_quantity)
            for gi in db.query(GrnItem).filter(GrnItem.grn_id == grn.id).all()
        ]
        result.append(GrnDTO(id=grn.id, poId=grn.po_id, items=items, receivedAt=grn.received_at))
    return result


def get_all_grns(db: Session) -> List[GrnDTO]:
    """Return every GRN across all Purchase Orders."""
    result = []
    for grn in db.query(Grn).all():
        items = [
            GrnItemDTO(itemId=gi.item_id, receivedQuantity=gi.received_quantity)
            for gi in db.query(GrnItem).filter(GrnItem.grn_id == grn.id).all()
        ]
        result.append(GrnDTO(id=grn.id, poId=grn.po_id, items=items, receivedAt=grn.received_at))
    return result
